### PR TITLE
[CPU] Add fused activation to Conv2D kernel implementation.

### PR DESCRIPTION
**Pull request by mciprian13**
**Summary**
- Added fused activation to CPU Conv2D kernel implementation.
- Added more fused activation types by adding an extra node/instruction member **FusedActivationArgs** for the activation parameters (e.g. min/max for Clip, alpha for LeakyRelu)
- Added **addFusedActivation()** utility in InstrBuilder.
- Propagated the fused activation where missing in some Conv2D transformations.
- Enabled a TFLite unit test which now works with the newest AvgPool implementation.

**Test Plan**
- Added unit tests in OperatorTest and GraphOptz.

### CODE REVIEW

**Comment by mciprian13** [comment]
@jfix71 Can you take a look on this? Thanks!

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 1]
```diff
@@ -4928,36 +4934,36 @@ template <class T> bool fuseActivation(T *N, Function *F, const Backend *B) {
...
+  case Kinded::Kind::LeakyReluNodeKind:
+    activationNV = cast<LeakyReluNode>(activation)->getResult();
+    N->setFusedActivation(FusedActivation::LEAKY_RELU);
+    N->setFusedActivationArgs({cast<LeakyReluNode>(activation)->getAlpha()});
     break;
   default:
     return false;
   }
 
-  // currently not to support asymmetric quantization fusion.
```
Comment (leejaymin) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:4950]: @mciprian13 Hello, I am Jemin. I wrote this code to prevent the fusion in case of asymmetric due to the offset. In asymmetric case, the offset of Conv. should be handled properly rather than replacing it by Relu's one. I wonder that you are aware of this issue.

**Review by leejaymin — COMMENTED** [review]
I have a quick question.

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp (hunk 1 above)
Comment (mciprian13) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:4950]: @leejaymin I don't understand what is the problem of fusing an asymmetric activation? The libjit correctly handles an activation with asymmetric quantization. You can look in libjit_defs.h at the function libjit_activation_q().

Comment (leejaymin) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:4950]: @mciprian13 Sorry for the ambiguous explanation. What I concern about is quantization parameters of convolution are substituted by ones of activation.  In the symmetric case, A(partialSum) * matMul(conv) / outScale (conv) * InScale(relu) / outSacle (relu) is equal to A * matmul(conv) / outScale(relu). It is beacuse outScale (conv) is equivalent to inScale(relu). Therefore, the replacement of quantization parameters is fine. However, In the asymmetric case, there are offsets in Conv and activation. Instead of replacing them with the following code `N->getResult().setType(activationNV.getType())`, the original offset of activation should be properly computed in libjit_activation_i32(). Let me know If I was wrong.

Comment (mciprian13) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:4950]: @leejaymin Changing the quantization parameters (scale and offset) of the convolution is not a problem because the mathematics remains consistent. Doing so actually has benefits because the dynamic range of activation output is commonly smaller than the dynamic range of the activation input (conv output). Computing the convolution using the quantization parameters of the activation:
- ensures better representation (smaller quantization step and hence error) for those values which are not saturated by Relu/Clip
- saturates those values which will be later saturated by Relu/Clip (don't care)
- this is the approach used by TensorFlowLite which ALWAYS uses fused activations in their pre-quantized models
From my experience fusing activations for quantized models always has benefits in terms of accuracy because it avoids an extra requantization (by using different quantization parameters at the interface between conv and activation).

**Comment by mciprian13** [comment]
@leejaymin Here is a doc describing why the math still works:
[Glow_Quantization.docx](https://github.com/pytorch/glow/files/5370528/Glow_Quantization.docx)

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp (hunk 1 above)
Comment (leejaymin) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:4950]: @mciprian13 Thank you for the really thorough answer and for sharing me the insightful document. As you explained in the attached document, I agreed with the benefit of performance due to mathematical simpleness. One thing I still concern about is that `S_y` and `O_y` are forced to match `S_a` and `O_a`. Such force conversion could affect the final accuracy. 
image
To figure out it, I tested accuracy variation by fusion on OpenCL backend and I got the following results:

```bash
# use imagenet12 validation set (50,000)
# baseline accuracy (FP32) top1 76.08%

# Resnet50, Asymmetric, Fusion (Conv and Relu)
Completed running; Final Top-1/5 accuracy across 50000 images:
       Top-1 accuracy: 0.11

# Resnet50, Asymmetric nonFusion
Completed running; Final Top-1/5 accuracy across 50000 images:
       Top-1 accuracy: 69.8
```

Even though the different thing between the two conditions is ON/OFF of Conv+Relu fusion, the accuracy drop is severe. However, except for the asymmetric case, the fusion slightly improves total accuracy as you mentioned. Therefore, I added the code to prevent fusion in the case of the asymmetric schema.
Currently, I am not sure that the poor accuracy stems from fusion or there might be something that is the wrong implementation in the OpenCL backend. 

In your code, I wonder if the force conversion doesn't matter.

Comment (mciprian13) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:4950]: @leejaymin The OpenCL implementation is obviously wrong:
https://github.com/pytorch/glow/blob/7fa10073a6641ecd53b3789acfe4d1c878155ce4/lib/Backends/OpenCL/kernels.cl#L1181
The max should be done with **destOffset** and not 0 to capture the general case for asymmetric quantization.

Comment (mciprian13) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:4950]: @leejaymin Could you try and compute again the accuracy? I fixed the OpenCL kernel by replacing 0 with destOffset.

Comment (leejaymin) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:4950]: @mciprian13 Thank you for your quick reply. 
I pulled your PR and ran it on CPU and OpenCL backends.
Even if you fixed the openCL code, the accuracy still shows the poor result as the below.
​```bash
Finished image index 1300 out of 50000
  Current Top-1/5 accuracy:
        Top-1 accuracy: 0.00
        Top-5 accuracy: 0.00
```
In CPU backend case, the fusion of the asymmetric case shows the right result.
​```bash
Finished image index 100 out of 50000
  Current Top-1/5 accuracy:
        Top-1 accuracy: 94.00
        Top-5 accuracy: 97.00
00:00:44.30
Finished image index 200 out of 50000
  Current Top-1/5 accuracy:
        Top-1 accuracy: 80.50
        Top-5 accuracy: 97.00
```
In light of the outcome, there is extra something wrong in the OpenCL backend.
Thank you for your clarification and guidance.
I hope your code is getting landed soon.

**Comment by mciprian13** [comment]
@jfix71 @jackm321 Is one of you available to review this? Thanks!

**Comment by mciprian13** [comment]
@842974287 Can you review this? Thanks!

**Review by jfix71 — COMMENTED** [review]
Cool @mciprian13 -- I am curious if you measured differences in perf here. Previously we were only ever fusing stacked elementwise ops IIRC. CC: @opti-mix

**Comment by mciprian13** [comment]
@jfix71 I measured the performance difference when compiling and running a floating-point MobileNet v1 0.25 128 (which is full of Clip nodes) on a ARM Cortex M7 core @ 600 MHz. The performance without fused activations was 212 ms while the performance with fused activations is 183 ms (I would say pretty significant).

File: tests/unittests/OperatorTest.cpp [hunk 2]
```diff
@@ -11639,6 +11639,111 @@ TEST_P(OperatorTest, NonSquareStrideConvolution) {
...
+  SaveNode *save = F->createSave("save", act);
+  auto *resultTensor = bindings.allocate(save->getPlaceholder());
+  return std::make_pair(F, resultTensor);
+}
+
+/// Check that Conv2D followed by activation works (whether fused or not).
+/// For this we compare with the Interpreter reference float implementation.
+#define TEST_CONV2D_ACTIVATION(ACTIVATION, TYPE, TOL)                          \
+  TEST_P(OperatorStatelessTest, Conv2D_##ACTIVATION##_##TYPE) {                \
+    ENABLED_BACKENDS("CPU");                                                   \
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply tests/unittests/OperatorTest.cpp:12125]: Not quite. When fusing the activations we actually get a better precision for the computation (that is because the activation is applied to the accumulator without using an extra requantization). The unit tests use error tolerances which reflect the better precision when using fused activations. When enabling these unit tests on Interpreter some of them fail because the Interpreter does not fuse the activations.

**Comment by mciprian13** [comment]
@jfix71 Any other items you want me to address?

File: lib/Backends/CPU/CPUBackend.cpp [hunk 3]
```diff
@@ -79,6 +79,39 @@ bool CPUBackend::shouldLower(const Node *N) const {
...
+    return true;
+  case Kinded::Kind::SigmoidNodeKind:
+    return actOutTy->isFPType();
+  case Kinded::Kind::TanhNodeKind:
+    return actOutTy->isFPType();
+  case Kinded::Kind::LeakyReluNodeKind:
+    return true;
+  default:
+    return false;
+  }
```
Comment (jfix71) [inline lib/Backends/CPU/CPUBackend.cpp:112]:
```suggestion
  // Only the following activations can be fused.
  // Additionally Tanh/Sigmoid are fused only for floating-point type. For
  // quantized type Lookup Tables should be used instead.
  switch (activation->getKind()) {
  case Kinded::Kind::ReluNodeKind:
  case Kinded::Kind::ClipNodeKind:
  case Kinded::Kind::LeakyReluNodeKind:
    return true;
  case Kinded::Kind::SigmoidNodeKind:
    return llvm::cast<SigmoidNode>(activation)->getResult().getType()->isFPType();
  case Kinded::Kind::TanhNodeKind:
    return llvm::cast<TanhNode>(activation)->getResult().getType()->isFPType();
  default:
    return false;
  }
```

File: tools/ClassGen/InstrBuilder.cpp [hunk 4]
```diff
@@ -558,3 +558,8 @@ InstrBuilder &InstrBuilder::addMember(MemberType type,
 
   return addMember(*typeInfo, name);
 }
+
+InstrBuilder &InstrBuilder::addFusedActivation() {
+  return addMember(MEMBER_TYPE_INFO(glow::FusedActivation), "FusedActivation")
```
Comment (jfix71) [inline tools/ClassGen/InstrBuilder.cpp:563]: Can you add a comment here explaining the need for `FusedActivationArgs`? E.g. something simple like some activations are parameterized, such as Clip which requires a min and max value.

File: lib/Backends/CPU/CPUBackend.cpp [hunk 5]
```diff
@@ -79,6 +79,39 @@ bool CPUBackend::shouldLower(const Node *N) const {
   }
 }
 
+bool CPUBackend::supportsFusedActivation(Node *parent, Node *activation) const {
+  // CPU backend only supports fusing activations into Convolution and
+  if (parent->getKind() != Kinded::Kind::ConvolutionNodeKind &&
+      parent->getKind() !=
+          Kinded::Kind::ChannelwiseQuantizedConvolutionNodeKind) {
+    return false;
+  }
```
Comment (jfix71) [inline lib/Backends/CPU/CPUBackend.cpp:88]:
```suggestion
  if (!llvm::isa<ConvolutionNode>(parent) &&
      !llvm::isa<ChannelwiseQuantizedConvolutionNode>(parent)) {
    return false;
  }
```

**Review by jfix71 — COMMENTED** [review]
Nice speedup!

I've added a few suggestions inside `supportsFusedActivation()`  -- I'd really prefer to avoid using `getNthResult()` unless absolutely necessary. Otherwise I think LG.

File: lib/Backends/CPU/CPUBackend.cpp [hunk 6]
```diff
@@ -79,6 +79,39 @@ bool CPUBackend::shouldLower(const Node *N) const {
   }
 }
 
+bool CPUBackend::supportsFusedActivation(Node *parent, Node *activation) const {
+  // CPU backend only supports fusing activations into Convolution and
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply lib/Backends/CPU/CPUBackend.cpp:83]: Done.

File: lib/Backends/CPU/CPUBackend.cpp (hunk 3 above)
Comment (mciprian13) [reply lib/Backends/CPU/CPUBackend.cpp]: Done.

File: lib/Backends/CPU/CPUBackend.cpp (hunk 5 above)
Comment (mciprian13) [reply lib/Backends/CPU/CPUBackend.cpp]: Done.

File: tools/ClassGen/InstrBuilder.cpp (hunk 4 above)
Comment (mciprian13) [reply tools/ClassGen/InstrBuilder.cpp:566]: Done.

**Comment by mciprian13** [comment]
@jfix71 Review items are done. Let's have this landed then 😄

**Review by jfix71 — APPROVED** [review]
Thanks!

**Comment by jfix71** [comment]
@mciprian13 FYI -- we saw significant perf degradation after this diff due to the CPU specialization for `CPUConvDKKC8Node` no longer applying. I'm wondering if you tested on our default CPU backend here, i.e. with `CPUConvDKKC8Node` included? I'm guessing you disabled this conversion when testing on an ARM core since I believe it was written somewhat specifically to an x86 core. Anyway, I think we should probably disable this fusion on the CPU backend until we can add the fusion support to the `CPUConvDKKC8` kernel so that we don't see this perf degradation. CC: @hl475

**Comment by mciprian13** [comment]
@jfix71 The performance of the `CPUConvDKKC8Node` on ARM is just awful. For small/medium sized models the default kernel is much better. Since the `CPUConvDKKC8Node` was written for x86 best approach would be to disable activation fusion and enable specialization to `CPUConvDKKC8Node` only for x86. For this we can use the backend's `getCPU` function to get the CPU name and do these 2 things conditionally.
